shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java

```diff
@@ -151,6 +159,44 @@ public void onApplicationEvent(final ContextRefreshedEvent contextRefreshedEvent
         }
     }
 
+    @Override
+    protected List<ApiDocRegisterDTO> buildApiDocDTO(final Class<?> clazz, final Method method) {
+        final String contextPath = getContextPath();
+        String apiDesc = Stream.of(method.getDeclaredAnnotations()).filter(item -> item instanceof ApiDoc).findAny().map(item -> {
+            ApiDoc apiDoc = (ApiDoc) item;
+            return apiDoc.desc();
+        }).orElse("");
+        String superPath = buildApiSuperPath(clazz, AnnotatedElementUtils.findMergedAnnotation(clazz, getAnnotationType()));
+        if (superPath.indexOf("*") > 0) {
+            superPath = superPath.substring(0, superPath.lastIndexOf("/"));
+        }
+        ShenyuSofaClient shenyuSofaClient = AnnotatedElementUtils.findMergedAnnotation(method, ShenyuSofaClient.class);
+        if (Objects.isNull(shenyuSofaClient)) {
+            return Lists.newArrayList();
+        }
+        List<ApiDocRegisterDTO> list = Lists.newArrayList();
+        String value = shenyuSofaClient.value();
+        String apiPath = contextPath + superPath + value;
+        ApiDocRegisterDTO build = ApiDocRegisterDTO.builder()
+                .consume(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .produce(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .httpMethod(ApiHttpMethodEnum.NOT_HTTP.getValue())
```

Comment:
I have seen a lot of similar code. Can you use template method pattern to abstract another parent class? The new abstract class extends from `AbstractContextRefreshedEventListener`, and `SofaServiceEventListener` extends from the new abstract class.

Reply:
i read and learn `template method`. and this code looks like `template method`.
```java
private void handleApiDoc(final String name, final Object bean) {
        Class<?> apiModuleClass = AopUtils.isAopProxy(bean) ? AopUtils.getTargetClass(bean) : bean.getClass();
        final Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(apiModuleClass);
        for (Method method : methods) {
            if (method.isAnnotationPresent(ApiDoc.class)) {
                List<ApiDocRegisterDTO> apis = buildApiDocDTO(bean, method);
                for (ApiDocRegisterDTO apiDocRegisterDTO : apis) {
                    publisher.publishEvent(apiDocRegisterDTO);
                }
            }
        }
    }
```
Is it necessary to add another abstract class?

Reply:
I think yes.

Reply:
> I think yes.

i will try

Reply:
> I think yes.

fix it,The template method is used to design the pattern, but I don't think it is necessary to add new subclasses. Do you think the current implementation method is OK?